File: array/0665_non_decreasing_array/0665_non_decreasing_array.py
# -*- coding: utf-8 -*-

class Solution(object):
    def checkPossibility(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        cnt = 0
        for i in range(1, len(nums)):
            if nums[i - 1] > nums[i]:
                cnt += 1
                if i < 2 or nums[i - 2] <= nums[i]:
                    nums[i - 1] = nums[i] # modify nums[i-1]
                else:
                    nums[i] = nums[i - 1] # modify nums[i]
        return cnt <= 1
            


print(Solution().checkPossibility([4,2,3]))

File: array/0665_non_decreasing_array/0665_non_decreasing_array.go
package main

import (
	"fmt"
)

func checkPossibility(nums []int) bool {
	if len(nums) == 0 || len(nums) == 1 || len(nums) == 2 {
		return true
	}

	found := false
	for i := 1; i < len(nums); i++ {
		if nums[i] < nums[i-1] {
			if found {
				return false
			}
			if i-2 >= 0 && nums[i] < nums[i-2] {
				nums[i] = nums[i-1]
			}
			found = true
		}
	}

	return true
}
func main() {
	nums := []int{4, 2, 3}
	fmt.Println(checkPossibility(nums))
}
